### PR TITLE
Fix #3851: Add metadata tab to raydata

### LINKED ISSUE #3851: raydata: metada tab
@ncook882 has created a notebook that shows how to access and visualize some metdata from bluesky. Need to pick the best way of accessing the data and add a tab to the app that show some visualizations of it.

@ncook882 says: 
The upshot is that there are three ways of accessing the data and associated metadata:
1. Databroker catalog structures - this is the most up-to-date and future-proofed representation, and the basis for a lot of the v2 API in Databroker. However, there is quite a bit of abstraction and hierarchy to them, so the learning curve may be steeper.
2. Databroker header structures - these are essential to the v1 API, and will be supported through backwards compatibility during v2. The header objects themselves are a bit flatter and bulkier than the catalog counterparts, so you get more immediate access to data, but they are more cumbersome to operate on.
3. Indirect access to structures via specific functions - The notebook we use leverages pyCHX functions to explore the metadata (via headers) and populate its own dictionary of metadata. Because this is part of the notebook workflow, we have to provide some environmental support for it. But its ultimately an obtuse and specific solution that (a) relies on others code, (b) doesn’t provide a lot of flexibility for the interface, and (c) doesn’t improve our understanding of databroker and how we can interact with data and metadata.

I’ve added a notebook for exploring metadata in a multitude of ways - https://github.com/radiasoft/raydata/blob/main/AnalysisNotebooks/CHX_three_components/XPCS_browse.ipynb

### PR DESCRIPTION
**Pull request by e-carlin**

### CODE REVIEW

File: sirepo/package_data/static/js/raydata.js [hunk 1]
```diff
@@ -41,10 +52,81 @@ SIREPO.app.directive('appHeader', function(appState, panelState) {
...
+                  </tbody>
+                </table>
+              </div>
+            </div>
+        `,
+        controller: function(appState, requestSender, $sce, $scope) {
+	    $scope.data = null;
+	    $scope.expanded = {};
+
+	    $scope.isOverflown = function(key) {
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/package_data/static/js/raydata.js:105]: I'm not sure about common usage. This is my first time doing something like this. I got the word from the css `text-overflow` property which adds the ellipsis when the text is long. Happy to use something else.

File: sirepo/package_data/static/js/raydata.js [hunk 2]
```diff
@@ -41,10 +52,81 @@ SIREPO.app.directive('appHeader', function(appState, panelState) {
...
+                    <td id="metadata-table-{{ k }}" class="raydata-overflow-text">{{ v }}</td>
+                    <td><button class="glyphicon glyphicon-plus" data-ng-if="isOverflown(k)" data-ng-click="toggleExpanded(k)"></span></td>
+                    <td><button class="glyphicon glyphicon-minus" data-ng-if="expanded[k]" data-ng-click="toggleExpanded(k)"></span></td>
+                  </tr>
+                  </tbody>
+                </table>
+              </div>
+            </div>
+        `,
+        controller: function(appState, requestSender, $sce, $scope) {
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/package_data/static/js/raydata.js]: 7355d2f3f

File: sirepo/package_data/static/js/raydata.js [hunk 3]
```diff
@@ -41,10 +52,81 @@ SIREPO.app.directive('appHeader', function(appState, panelState) {
...
+	    $scope.data = null;
+	    $scope.expanded = {};
+
+	    $scope.isOverflown = function(key) {
+		// Id's may be malformed (spaces) so use this jquery syntax
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/package_data/static/js/raydata.js]: 9a9f3cfca

File: sirepo/package_data/static/js/raydata.js [hunk 4]
```diff
@@ -41,10 +52,81 @@ SIREPO.app.directive('appHeader', function(appState, panelState) {
...
+		// to cover them
+		const e = $(`[id='metadata-table-${key}']`);
+		return e.prop('clientWidth') < e.prop('scrollWidth');
+	    };
+
+	    $scope.toggleExpanded = function(key) {
+		const e = $(`[id='metadata-table-${key}']`);
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/package_data/static/js/raydata.js]: 9a9f3cfca

File: sirepo/package_data/static/js/raydata.js [hunk 5]
```diff
@@ -41,10 +52,81 @@ SIREPO.app.directive('appHeader', function(appState, panelState) {
...
+		if (!(key in $scope.expanded)) {
+		    $scope.expanded[key] = false;
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/package_data/static/js/raydata.js]: 0afd2ac82 I'm surprised you find this cleaner. My general rule of thumb for writing code you like is "avoid else when possible". 

I find the way I had it written cleaner. It uses a `!` which is not so good but I like that it has a guard clause for the weird case (uninitialized) and then all of the normal case code is right next to each other.

File: sirepo/package_data/static/js/raydata.js [hunk 6]
```diff
@@ -41,10 +52,81 @@ SIREPO.app.directive('appHeader', function(appState, panelState) {
...
+		}
+		$scope.expanded[key] = ! $scope.expanded[key];
+		e.toggleClass('raydata-overflow-text');
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/package_data/static/js/raydata.js]: I don't see the upside to that under our current model. The truth is that this table should be generalized across apps (at the very least shared with jspec) and then sharing the css would make more sense. I never know how to prioritize refactors like this vs "getting things done". If you like I'd be happy to generalize our table code and share the css.

File: sirepo/package_data/static/js/raydata.js [hunk 7]
```diff
@@ -41,10 +52,81 @@ SIREPO.app.directive('appHeader', function(appState, panelState) {
...
+	    };
+
+	    requestSender.statelessCompute(
+		appState,
+		{
+		    method: $scope.type + '_metadata'
+		},
+		(data) => {
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/package_data/static/js/raydata.js:143]: No, edcbcd62f . Better?

[the remark this answers is not in the document]
Comment (robnagler) [reply sirepo/package_data/static/js/raydata.js:143]: Better in that it handles errors. Not good in that we should default statelessCompute to handle errors

File: sirepo/template/raydata.py [hunk 8]
```diff
@@ -6,22 +6,66 @@
...
+)
+
 
 def background_percent_complete(report, run_dir, is_running):
     if is_running:
         return PKDict(percentComplete=0, frameCount=0)
     return PKDict(percentComplete=100, frameCount=1)
 
 
+def stateless_compute_analysis_metadata(data):
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/template/raydata.py:57]: Not really sure what you mean wrt override dispatch. Something like this?
```python
def stateless_compute(method, data):
  return PKDict(data=_metadata(PKDict(analysis=_ANALAYSIS_METADATA, ...)[method])
```

File: sirepo/package_data/static/js/raydata.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (robnagler) [reply sirepo/package_data/static/js/raydata.js:123]: it was the "isOverflow<n>". How about isOverflowing or wouldOverflow?

File: sirepo/template/raydata.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (robnagler) [reply sirepo/template/raydata.py:57]: The background_percent_complete should be some type of computation or default to the above in the superclass (which doesn't exist).

File: sirepo/template/raydata.py [hunk 9]
```diff
@@ -6,22 +6,66 @@
...
+    'plan_args',
+    'plan_name',
+    'plan_type',
+    'scan_id',
+    'sequence id',
+)
+
 
 def background_percent_complete(report, run_dir, is_running):
     if is_running:
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/template/raydata.py:53]: I'd be happy to brainstorm ideas and implement something.

File: sirepo/template/raydata.py [hunk 10]
```diff
@@ -6,22 +6,66 @@
...
+
+# TODO(e-carlin): from user
+_SCAN_UID = 'bdcce1f3-7317-4775-bc26-ece8f0612758'
+
 _OUTPUT_FILE = 'out.ipynb'
 
+_GENERAL_METADATA = (
+    'beamline_id',
+    'cycle',
+    'data path',
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/template/raydata.py:37]: 9f7d1e470

File: sirepo/template/raydata.py [hunk 11]
```diff
@@ -6,22 +6,66 @@
...
+    'time',
+    'uid',
+)
+
+_PLAN_METADATA = (
+    'plan_args',
+    'plan_name',
+    'plan_type',
+    'scan_id',
+    'sequence id',
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/template/raydata.py:48]: 9f7d1e470

File: sirepo/template/raydata.py [hunk 12]
```diff
@@ -37,3 +81,12 @@ def _generate_parameters_file(data):
             output_name=_OUTPUT_FILE,
         ),
     )
+
+
+def _metadata(type):
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/template/raydata.py]: 9f7d1e470

File: sirepo/package_data/static/js/raydata.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (robnagler) [reply sirepo/package_data/static/js/raydata.js]: It was the assigning it false and then inverting that to true. The default is true so it's easier to read this way.

"else" is removed when the assignment makes sense, e.g. 

```
m = someComputedValue()
if m is None:
    m = default_thing
```

In this case "m" is complex to compute so we do that once, and then override with a default if it isn't there or whatever.

File: sirepo/package_data/static/js/raydata.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/package_data/static/js/raydata.js]: 607dfa6fb

File: sirepo/package_data/static/js/raydata.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/package_data/static/js/raydata.js:144]: Agreed. Let's talk about what you'd like to see.

File: sirepo/template/raydata.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/template/raydata.py:58]: > "ANALAYSIS_METADATA" should be looked a programmatic control based on method

Mind saying that a different way? I'm unfamiliar with the term "programmatic control".

**Comment by e-carlin** [comment]
GitHub thinks I am still waiting on your review. I think not. So, requesting another review here...

File: sirepo/package_data/static/js/raydata.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/package_data/static/js/raydata.js]: Thanks.

File: sirepo/package_data/static/js/raydata.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/package_data/static/js/raydata.js]: We talked - decided to save this for a later time.

File: sirepo/package_data/static/js/raydata.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/package_data/static/js/raydata.js:145]: 39b3325eb

File: sirepo/template/raydata.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/template/raydata.py]: 2dfe007af

File: sirepo/package_data/static/css/raydata.css [hunk 13]
```diff
@@ -0,0 +1,6 @@
+.raydata-overflow-text {
```
Comment (e-carlin) [inline sirepo/package_data/static/css/raydata.css:1]: @moellep If you have better suggestions for css please let me know. The most interesting case is when you expand the plan args field in the plan report and then adjust the browser window in the horizontal direction. In some cases a scroll bar is added so the content doesn't overflow the report panel. We don't use scrollbars in many places so if you'd like something different lmk.

**Comment by e-carlin** [comment]
@robnagler ready for another review.

**Review by robnagler — APPROVED** [review]

**Pull request merged**